### PR TITLE
Switch to scala 2.11

### PR DESCRIPTION
**Pull request by davidegrohmann**
Refactor the code generation part of the new runtime in order to avoid inheritance across Java and Scala which causes compilation errors.

### CODE REVIEW

File: community/cypher/cypher/pom.xml [hunk 1]
```diff
@@ -169,8 +169,9 @@
     <!-- neo4j-cypher -->
     <dependency>
       <groupId>org.neo4j</groupId>
-      <artifactId>neo4j-cypher-compiler-1.9</artifactId>
-      <version>2.0.4</version>
+      <artifactId>neo4j-cypher-compiler-1.9_2.11</artifactId>
+      <!-- FIXME: this should be replaced by the 2.0.5 released 1.9 compiler -->
+      <version>2.0.5-SNAPSHOT</version>
```
Comment (davidegrohmann) [inline community/cypher/cypher/pom.xml]: Replace with a stable release when ready.

**Comment by davidegrohmann** [comment]
@thobe  I had to restructure a bit the `CompiledExecutionResult` used by the new runtime and the code generation as well since having a Java class extending a Scala class was causing compilation errors. The current solution is probably sub-optimal, but perhaps good enough to allow us for merging this PR in and cleanup later. WDYT?

File: community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ToCompileCode.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.executionplan;
+
+import java.util.List;
+
+import org.neo4j.cypher.internal.compiler.v2_3.ExecutionMode;
+import org.neo4j.cypher.internal.compiler.v2_3.planDescription.InternalPlanDescription;
+import org.neo4j.graphdb.Result;
+
+public interface ToCompileCode
```
Comment (thobe) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ToCompileCode.java]: The name of this interface is awful!

**Comment by thobe** [comment]
@davidegrohmann The change to have a simple interface implemented by the generated code and then a wrapper to add the support that we want for the tests that are written in scala is sensible, it is what we've been thinking we should do eventually anyhow. We would want to move towards the interface implemented by the generated code being the interface expected by the APIs in kernel, so that there is no indirection in the common case (in the tests in scala we can afford the indirection, but when actually executing queries we would prefer to be without it), but this is something we will refactor towards later, so it's acceptable that this is not done yet as part of this commit, we can fix that later. The name of the interface you introduced is absolutely horrible though, which makes me think that perhaps we should just use the correct interface now, to avoid having to name this short-lived interface.

File: community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ToCompileCode.java (hunk 2 above)
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ToCompileCode.java]: I agree, not proud of that naming. If you have any good suggestion please let me know.

Comment (thobe) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ToCompileCode.java]: `GeneratedQueryExecution` or something like that...

**Comment by pontusmelke** [comment]
I think it would like to change so that the generated java class only extends pure java stuff, but that is for a later day. However you have lots of licenses fun left…

**Comment by unoexperto** [comment]
Hey guys,

Any ETA on implementing this ?

Thanks.

**Comment by jotomo** [comment]
@cppexpert The latest milestone (2.3.0-M02) is using Scala 2.11 already :-)

**Comment by unoexperto** [comment]
My god ! I'm so going to try it now. And when is release scheduled ?

**Pull request merged**